sqlglot/optimizer/qualify_columns.py

```diff
@@ -528,61 +529,78 @@ def get_table(self, column_name: str) -> t.Optional[exp.Identifier]:
         return exp.to_identifier(table_name)
 
     @property
-    def all_columns(self):
+    def all_columns(self) -> t.Set[str]:
         """All available columns of all sources in this scope"""
         if self._all_columns is None:
             self._all_columns = {
                 column for columns in self._get_all_source_columns().values() for column in columns
             }
         return self._all_columns
 
-    def get_source_columns(self, name, only_visible=False):
-        """Resolve the source columns for a given source `name`"""
+    def get_source_columns(self, name: str, only_visible: bool = False) -> t.List[str]:
+        """Resolve the source columns for a given source `name`."""
         if name not in self.scope.sources:
             raise OptimizeError(f"Unknown table: {name}")
 
-        source = self.scope.sources[name]
-
-        # If referencing a table, return the columns from the schema
-        if isinstance(source, exp.Table):
-            return self.schema.column_names(source, only_visible)
+        original_source = self.scope.sources[name]
 
-        if isinstance(source, Scope) and isinstance(source.expression, exp.Values):
-            return source.expression.alias_column_names
+        if isinstance(original_source, exp.Table):
+            # If referencing a table, return the columns from the schema
+            columns = self.schema.column_names(original_source, only_visible)
+        elif isinstance(original_source, Scope) and isinstance(
+            original_source.expression, exp.Values
+        ):
+            columns = original_source.expression.alias_column_names
+        else:
+            # Otherwise, if referencing another scope, return that scope's named selects
+            columns = original_source.expression.named_selects
+
+        source = seq_get(self.scope.selected_sources.get(name) or [], 0)
```

Comment:
let's just do

source = self.scope.selected_sources[name] if name in ... else None

Reply:
Refactored it, a little bit differently but imo clear too. Happy to change later if you prefer this form.